onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+      }
+    }
+    if (cast_fp16_siblings.size() > 1) {
+      modified = true;
+      FuseNodes(graph, output, cast_fp16_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp16_siblings);
+    }
+    if (cast_fp32_siblings.size() > 1) {
+      modified = true;
+      FuseNodes(graph, output, cast_fp32_siblings, removed_nodes);
+      VLOGS(logger, 1) << "FusedSubgraphs: Fused Cast nodes : " << ConcatNames<std::vector<Node*>>(cast_fp32_siblings);
+    }
+  }
+  return modified;
+}
+
+// RemoveUnnecessaryCasts
+// Remove a cast if the input elem_type is same the required cast type.
+static bool RemoveUnnecessaryCasts(Graph& graph, Node* node,
+                                   std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                   const logging::Logger& logger) {
+  bool modified = false;
+  if (node->InputDefs().size() == 1) {
+    const NodeArg* node_arg = node->InputDefs()[0];
+    auto elem_type = node_arg->TypeAsProto()->tensor_type().elem_type();
+    TensorProto_DataType data_type = static_cast<TensorProto_DataType>(elem_type);
+    if (IsCastTo(node, data_type)) {
+      VLOGS(logger, 1) << "Removed unnecessary cast " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// PropagateFP32CastsFromInputsToOutputs
+// This non recursive fusion, checks whether the given node is fp16 safe op and
+// whether all floatingpoint inputs are cast to fp32
+// and propagates cast op to the floatingpoint outputs.
+static bool PropagateFP32CastsFromInputsToOutputs(Graph& graph, Node* node,
+                                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                                  size_t level,
+                                                  const logging::Logger& logger) {
+  bool modified = false;
+  if (IsFP16Allow(node->OpType(), level)) {
+    bool has_float_inputs = false;
+    bool all_float_inputs_have_casts = true;
+    std::vector<Node*> casts;
+    std::unordered_set<NodeArg*> require_type_change;
+    for (NodeArg* input : node->MutableInputDefs()) {
+      if (!IsType(*input, TensorProto::FLOAT)) {
+        continue;
+      }
+      has_float_inputs = true;
+      Node* producer = graph.GetMutableProducerNode(input->Name());
+      if (producer && IsCastTo(producer, TensorProto::FLOAT)) {
+        casts.push_back(producer);
+        require_type_change.insert(input);
+        continue;
+      }
+      all_float_inputs_have_casts = false;
+      break;
+    }
+    if (has_float_inputs && all_float_inputs_have_casts && casts.size() > 1) {
+      VLOGS(logger, 1) << "PropagateFP32CastsFromInputsToOutputs: Removed Cast nodes "
+                       << ConcatNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
```

Comment:
>node->Name() [](start = 63, length = 12)

In this log, let's make it explicit that "node->Name()" has been place in fp16.

Reply:
The PropagateBackward/PropagateForward functions already  print the NodeArg names that changed the type.

Reply:
NodeArg will have the names like "123", "234"... 
Node name will tell us more about the which ops are placed in fp16. 
commonly, they have the name like Gemm_123